java/ql/lib/semmle/code/java/security/InsufficientKeySize.qll

```diff
@@ -0,0 +1,191 @@
+          DataFlow::localExprFlow(kpg, kpgInit.getQualifier()) and
+          this.asExpr() = kpgInit.getKeySizeArg()
+        )
+        or
+        exists(Spec spec | this.asExpr() = spec.getKeySizeArg())
+      }
+
+      override predicate hasState(DataFlow::FlowState state) { state = getMinKeySize().toString() }
+    }
+
+    /** Returns the minimum recommended key size for RSA, DSA, and DH algorithms. */
+    private int getMinKeySize() { result = 2048 }
+
+    /** An instance of an RSA, DSA, or DH algorithm specification. */
+    private class Spec extends ClassInstanceExpr {
+      Spec() {
+        this.getConstructedType() instanceof RsaKeyGenParameterSpec or
+        this.getConstructedType() instanceof DsaGenParameterSpec or
+        this.getConstructedType() instanceof DhGenParameterSpec
+      }
+
+      /** Gets the `keysize` argument of this instance. */
+      Argument getKeySizeArg() { result = this.getArgument(0) }
+    }
+  }
+
+  /** Provides models for elliptic-curve asymmetric cryptography. */
+  private module EllipticCurve {
+    /** A source for an insufficient key size used in elliptic curve (EC) algorithms. */
+    private class Source extends InsufficientKeySizeSource {
+      Source() {
+        this.asExpr().(IntegerLiteral).getIntValue() < getMinKeySize()
+        or
+        // the below is needed for cases when the key size is embedded in the curve name
+        getKeySize(this.asExpr().(StringLiteral).getValue()) < getMinKeySize()
+      }
+
+      override predicate hasState(DataFlow::FlowState state) { state = getMinKeySize().toString() }
+    }
+
+    /** A sink for an insufficient key size used in elliptic curve (EC) algorithms. */
+    private class Sink extends InsufficientKeySizeSink {
+      Sink() {
+        exists(KeyPairGenInit kpgInit, KeyPairGen kpg |
+          kpg.getAlgoName().matches("EC%") and
+          DataFlow::localExprFlow(kpg, kpgInit.getQualifier()) and
+          this.asExpr() = kpgInit.getKeySizeArg()
+        )
+        or
+        exists(Spec s | this.asExpr() = s.getKeySizeArg())
+      }
+
+      override predicate hasState(DataFlow::FlowState state) { state = getMinKeySize().toString() }
+    }
+
+    /** Returns the minimum recommended key size for elliptic curve (EC) algorithms. */
+    private int getMinKeySize() { result = 256 }
+
+    /** Returns the key size from an EC algorithm's curve name string */
+    bindingset[algorithm]
+    private int getKeySize(string algorithm) {
+      algorithm.matches("sec%") and // specification such as "secp256r1"
+      result = algorithm.regexpCapture("sec[p|t](\\d+)[a-zA-Z].*", 1).toInt()
+      or
+      algorithm.matches("X9.62%") and //specification such as "X9.62 prime192v2"
+      result = algorithm.regexpCapture("X9\\.62 .*[a-zA-Z](\\d+)[a-zA-Z].*", 1).toInt()
+      or
+      (algorithm.matches("prime%") or algorithm.matches("c2tnb%")) and //specification such as "prime192v2"
```

Comment:
Simplification:

```suggestion
      algorithm.matches(["prime%", "c2tnb%"]) and //specification such as "prime192v2"
```

Reply:
Updated in https://github.com/github/codeql/pull/10785/commits/65f74741103383fdd8e16ebb2c5e4d248e0e7c6c.